java/arcs/core/analysis/Deduction.kt

```diff
@@ -0,0 +1,108 @@
+package arcs.core.analysis
+
+import arcs.core.data.expression.Expression
+
+/** Field [Identifier]. Lists of [Identifier]s imply an AccessPath.*/
+typealias Identifier = String
+
+/**
+ * [Deduction]s represent how inputs of a Paxel [Expression] contribute to its outputs.
+ *
+ * [Deduction]s recursively associate identifiers to field paths. These relationships can be
+ * directly translated into [Claim]s and [AccessPath]s.
+ */
+sealed class Deduction {
+
+  /** Substitute aliases in all paths. */
+  abstract fun substitute(aliases: Scope): Deduction
+
+  /** Union of two [Deduction] collections. */
+  abstract operator fun plus(other: Deduction): Deduction
+
+  /** A representation of a lookup claim in a Paxel [Expression]. */
+  data class Equal(val path: List<Identifier> = emptyList()) : Deduction() {
+
+    constructor(vararg paths: Identifier) : this(paths.toList())
+
+    /** Return [Equal] with all alias substitutions applied. */
+    override fun substitute(aliases: Scope): Equal = Equal(
+      path.flatMap { identifier ->
+        when (val association = aliases.associations.getOrDefault(identifier, Equal(identifier))) {
+          is Equal -> association.path
+          is Derive -> association.firstPath().path
+          is Scope -> throw UnsupportedOperationException("Cannot substitute a Scope for an Identifier.")
+        }
+      }
+    )
+
+    /** Append more [Identifier]s to the [Equal] */
+    fun mergeTop(other: Equal) = Equal(path + other.path)
+
+    /** Union of a [Equal] and a [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Equal -> Derive(this) + Derive(other)
+      is Derive -> Derive(this) + other
+      is Scope -> Derive(this) + Derive(other)
+    }
+  }
+
+  /** A representation of a Derivation claim in a Paxel [Expression]. */
+  data class Derive(val paths: List<Deduction> = emptyList()) : Deduction() {
```

Comment:
as commented below, I think derive should contain a list of paths, not a list of deductions. Do you have an example where derive composing deductions is useful?